### PR TITLE
start magicgui apps too

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
it was discovered in https://github.com/napari/magicgui/issues/54 that creating a magicgui widget before then using `napari.gui_qt` will prevent the application from running at all.  For instance, if you run the following script from the command line, it will _not_ show the viewer... it will just seem to do nothing.

```python
import napari
from magicgui import magicgui

@magicgui
def test(x: int):
    pass

with napari.gui_qt():
    v = napari.Viewer()
```

The reason is that magicgui has instantiated the (global) Qt app when the decorator is called, but the `napari.gui_qt` context manager _only_ starts applications that are named `"napari"` (with the intention of avoiding unnecessary blocking in ipython):
https://github.com/napari/napari/blob/1d784cf8373495d9591594b6dd6ac479d5566ed1/napari/_qt/event_loop.py#L86-L89

I'm wondering if you'd consider the following PR to allow napari to start magicgui-created applications.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/_qt/event_loop.py
Comment (sofroniewn) [inline napari/_qt/event_loop.py]: Do you know how long this import takes? We put a lot of work during #758 to reducing imports/ making lazy to improve startup time, and it looks like this would run everytime we launch even if not using magicgui. (Not sure if we're already importing magicgui here too https://github.com/napari/napari/blob/master/napari/utils/_magicgui.py and so maybe this doesn't change anything, but I thought I should ask).


The alternative I see is just hardcoding it to `magicgui`, and if that ever changes then we need to make an update.

Thoughts @jni @tlambert03?

**Comment by jni** [comment]
I don't have any specific problems with this as a bandaid, but it seems a bit like kicking the can down the road... What if someone comes along and produces an even better mugglegui? And people want to start using it in napari?

@sofroniewn re your very valid concern about import times, now that we are 3.7+ only, lazy imports become pretty easy to do, and I think we should start making use of it pretty indiscriminately 😂. See this PR: https://github.com/scikit-image/scikit-image/pull/5101

File: napari/_qt/event_loop.py
Comment (tlambert03) [reply napari/_qt/event_loop.py]: > Do you know how long this import takes? We put a lot of work during #758 to reducing imports/ making lazy to improve startup time, and it looks like this would run everytime we launch even if not using magicgui. 

takes about 15ms on my computer... I'm totally fine with making it a string though.

**Comment by tlambert03** [comment]
> but it seems a bit like kicking the can down the road... What if someone comes along and produces an even better mugglegui

:joy: ... love the name.  Yep it's possible, but it feels extremely unlikely that that list will ever get more than a few more.  plus, I can't think of an alternative approach... any ideas?

**Comment by tlambert03** [comment]
> I can't think of an alternative approach... any ideas?

I guess I can think of another: napari provides a "hook" (not an actual hookspec though) `start_my_application(name: str)` that magicgui could call to add its name to the list

File: napari/_qt/event_loop.py
Comment (sofroniewn) [reply napari/_qt/event_loop.py]: 15ms isn't too bad.

**Comment by sofroniewn** [comment]
> I don't have any specific problems with this as a bandaid, but it seems a bit like kicking the can down the road... What if someone comes along and produces an even better mugglegui? And people want to start using it in napari?

I think i'm ok with a bandaid for 2, where the other is something we maintain. I think if we find 3 then maybe we should consider a more generalizable fix

File: napari/_qt/event_loop.py
Comment (sofroniewn) [reply napari/_qt/event_loop.py]: I'd still vote for string, but final decision can be @tlambert03 and @jni

Comment (tlambert03) [reply napari/_qt/event_loop.py]: let's go with string.  It's already a bandaid 😂

Comment (tlambert03) [reply napari/_qt/event_loop.py]: there.  now it's a one-liner

**Comment by jni** [comment]
Three half-baked ideas that probably betray my brittle understanding of this issue:

1. add a keyword argument to provide "valid" application names, defaults to `['napari']` but can be set to anything.
2. add an argument `any=False` that can be set to True to ignore the name check
3. add an argument to *magicgui* to set the application name, and the above becomes `@magicgui(application='napari')`

?

File: napari/_qt/event_loop.py
Comment (sofroniewn) [inline napari/_qt/event_loop.py:89]: Not to be be nitpicky, but do you want to keep
```python
COMPATIBLE_APPLICATIONS = (`napari`, `magicgui`)
```
and then make this `in COMPATIBLE_APPLICATIONS`. Might be a little cleaner more/ future proof

Comment (jni) [reply napari/_qt/event_loop.py:89]: I don't mind the new string approach, I thought it was quite heavyweight above...

Comment (jni) [reply napari/_qt/event_loop.py:89]: BUT, there should be a comment here linking to this PR and explaining that we treat magicgui applications specially because the decorator creates them even before napari has had a chance to do so.

Comment (sofroniewn) [reply napari/_qt/event_loop.py:89]: This would just be variable, not two like before. It would then be very easy to associate comment with that variable when it's defined and not in the middle of event loop code, but i'll defer to @tlambert03

Comment (tlambert03) [reply napari/_qt/event_loop.py:89]: > * add a keyword argument to provide "valid" application names, defaults to `['napari']` but can be set to anything.
> * add an argument `any=False` that can be set to True to ignore the name check
> * add an argument to _magicgui_ to set the application name, and the above becomes `@magicgui(application='napari')`

these would all "work" ... but I feel like we're passing the burden of understanding what's happening here to users.  I can see "wait, when do I need to use `magicgui(application=napari)`?" (answer: only if you use the decorator _before_ calling `napari.gui_qt`), "when do I need to use `any=False`" (answer: be careful not to use it in a jupyter environment or you'll get blocking!" ...
while I agree this isn't the most elegant thing, it feels like the lesser of two weevils.

**Review by jni — APPROVED** [review]
@tlambert03 ok let's move forward with this, but please add that comment linking to this PR so we can find the discussion in the future! Merge when happy. =)

**Comment by tlambert03** [comment]
added.  thanks both!

**Review by sofroniewn — APPROVED** [review]
Comment has been added, I'm fine with current usage of strings. I will merge once tests have all passed

**Comment by sofroniewn** [comment]
Looks like it's in!

**Comment by tlambert03** [comment]
hah.  sorry! :) I felt (perhaps too) certain that it wouldn't affect the tests (famous last words)

**Pull request merged**
